comp/networkpath/npscheduler/npschedulerimpl/npscheduler.go

```diff
@@ -7,31 +7,324 @@
+	}
+}
+func (s *npSchedulerImpl) start() error {
+	if s.running {
+		return errors.New("server already started")
+	}
+	s.running = true
+	s.logger.Info("Start NpScheduler")
+	go s.listenPathtests()
+	go s.flushLoop()
+	s.startWorkers()
+	return nil
 }
 
-func newNpSchedulerImpl(epForwarder eventplatform.Component) *npSchedulerImpl {
-	return &npSchedulerImpl{
-		enabled:     true,
-		epForwarder: epForwarder,
+func (s *npSchedulerImpl) stop() {
+	s.logger.Info("Stop NpScheduler")
+	if !s.running {
+		return
+	}
+	close(s.stopChan)
+	<-s.flushLoopDone
+	<-s.runDone
+	s.running = false
+}
+
+func (s *npSchedulerImpl) listenPathtests() {
+	s.logger.Debug("Starting listening for pathtests")
+	for {
+		select {
+		case <-s.stopChan:
+			s.logger.Info("Stopped listening for pathtests")
+			s.runDone <- struct{}{}
+			return
+		case ptest := <-s.pathtestInputChan:
+			s.logger.Debugf("Pathtest received: %+v", ptest)
+			s.receivedPathtestCount.Inc()
+			s.pathtestStore.Add(ptest)
+		}
+	}
+}
+
+func (s *npSchedulerImpl) runTracerouteForPath(ptest *pathteststore.PathtestContext) {
+	s.logger.Debugf("Run Traceroute for ptest: %+v", ptest)
+
+	startTime := time.Now()
+	cfg := traceroute.Config{
+		DestHostname: ptest.Pathtest.Hostname,
+		DestPort:     ptest.Pathtest.Port,
+		MaxTTL:       0, // TODO: make it configurable, setting 0 to use default value for now
+		TimeoutMs:    0, // TODO: make it configurable, setting 0 to use default value for now
+	}
+
+	path, err := s.runTraceroute(cfg)
+	if err != nil {
+		s.logger.Errorf("%s", err)
+		return
+	}
+
+	s.sendTelemetry(path, startTime, ptest)
+
+	payloadBytes, err := json.Marshal(path)
+	if err != nil {
+		s.logger.Errorf("json marshall error: %s", err)
+	} else {
+		s.logger.Debugf("network path event: %s", string(payloadBytes))
+		m := message.NewMessage(payloadBytes, nil, "", 0)
+		err = s.epForwarder.SendEventPlatformEventBlocking(m, eventplatform.EventTypeNetworkPath)
+		if err != nil {
+			s.logger.Errorf("failed to send event to epForwarder: %s", err)
+		}
+	}
+}
+
+func runTraceroute(cfg traceroute.Config) (payload.NetworkPath, error) {
+	tr, err := traceroute.New(cfg)
+	if err != nil {
+		return payload.NetworkPath{}, fmt.Errorf("new traceroute error: %s", err)
+	}
+	path, err := tr.Run(context.TODO())
+	if err != nil {
+		return payload.NetworkPath{}, fmt.Errorf("run traceroute error: %s", err)
+	}
+	return path, nil
+}
+
+func (s *npSchedulerImpl) flushLoop() {
+	s.logger.Debugf("Starting flush loop")
+	defer s.logger.Debugf("Stopped flush loop")
+
+	flushTicker := time.NewTicker(s.flushInterval)
+
+	var lastFlushTime time.Time
+	for {
+		select {
+		// stop sequence
+		case <-s.stopChan:
+			s.logger.Info("Stop flush loop")
+			s.flushLoopDone <- struct{}{}
+			flushTicker.Stop()
+			return
+		// automatic flush sequence
+		case now := <-flushTicker.C:
+			s.flushWrapper(now, lastFlushTime)
+		}
+	}
+}
+
+func (s *npSchedulerImpl) flushWrapper(now time.Time, lastFlushTime time.Time) {
```

Comment:
Should you use `s.TimeNowFn` for `now` to be consistent?

Reply:
I update this one here: d61dfa2f33 

Do you see more cases?

Reply:
I was trying to point out the case here where `now` may not be the same as what we get from `s.TimeNowFn` (e.g., in a mocked test environment). Would just not passing in `now` here and using the value from `s.TimeNowFn` in the function work here?

Reply:
Oh, I see, for `flushWrapper()` I think `now` argument might be confusing. I think in this case, the "now" refers to the flush time. So maybe better to rename the variable to make it clear.

I just renamed from `now` to `flushTime` instead, which makes more sense in this context.

Fixed here 174f5dc661

Reply:
So even now I think you are doing a `Sub` with `TimeNowFn` and `flushTime`. If `TimeNowFn` was mocked in a test, can an operation with these two timestamps like `Sub` still be valid?

Reply:
@hmahmood I managed to make TimeNowFn to return something different at every call like this, but still being predictible: https://github.com/DataDog/datadog-agent/blob/4c091078a315c8b22cb0a4e1cf1663f3fc3c667f/comp/networkpath/npscheduler/npschedulerimpl/npscheduler_test.go#L255-L260

Reply:
I guess it is fine for now, but if you do add a test later that does involve the ticker it may not behave as expected. 

What I was trying to suggest was to call `flushWrapper(s.TimeNowFn(), ...)` instead of passing in the time value from the ticker.